### ISSUE TITLE
 `nonexistent-origin` exists in GNU/Linux

### ISSUE THREAD

**Issue by jugglinmike** [w3c/web-platform-tests#10560]
[The documentation on WPT server features](https://github.com/w3c/web-platform-tests/blob/0628e2e73489ddd075f273d2555e504e97e28ce6/docs/_writing-tests/server-features.md) states:

> In the test environment, five subdomains are available: `www`, `www1`, `www2`, `天気の良い日`, and `élève`; there is also `nonexistent-origin` which is guaranteed not to resolve.

However, following the instructions for system configuration causes requests to the domain `nonexistent-origin.web-platform.test` to resolve to `web-platform.test`:

    $ curl nonexistent-origin.web-platform.test:8000 -o /dev/null -sv
    * Rebuilt URL to: nonexistent-origin.web-platform.test:8000/
    *   Trying 0.0.0.0...
    * Connected to nonexistent-origin.web-platform.test (127.0.0.1) port 8000 (#0)
    > GET / HTTP/1.1
    > Host: nonexistent-origin.web-platform.test:8000
    > User-Agent: curl/7.47.0
    > Accept: */*
    > 
    < HTTP/1.1 200 OK
    < Content-Type: text/html
    < Server: BaseHTTP/0.3 Python/2.7.12
    < Date: Sat, 21 Apr 2018 00:07:47 GMT
    < Content-Length: 11853
    < 
    { [8192 bytes data]
    * Connection #0 to host nonexistent-origin.web-platform.test left intact

This is because the `./wpt create-hosts-file` command includes the following directive:

    0.0.0.0 nonexistent-origin.web-platform.test

In GNU/Linux, the address 0.0.0.0 is interpreted to mean "any IP address," so the WPT server (bound to 127.0.0.1) responds to requests for this address:

    $ curl 0.0.0.0:8000 -o /dev/null -sv
    * Rebuilt URL to: 0.0.0.0:8000/
    *   Trying 0.0.0.0...
    * Connected to 0.0.0.0 (127.0.0.1) port 8000 (#0)
    > GET / HTTP/1.1
    > Host: 0.0.0.0:8000
    > User-Agent: curl/7.47.0
    > Accept: */*
    > 
    < HTTP/1.1 200 OK
    < Content-Type: text/html
    < Server: BaseHTTP/0.3 Python/2.7.12
    < Date: Sat, 21 Apr 2018 00:44:31 GMT
    < Content-Length: 11853
    < 
    { [8192 bytes data]
    * Connection #0 to host 0.0.0.0 left intact

However, in Windows desktop environments, the same address is interpreted as non-existent:

    PS C:\Users\admin> Invoke-RestMethod -Uri 0.0.0.0:8000
    Invoke-RestMethod : Unable to connect to the remote server
    At line:1 char:1
    + Invoke-RestMethod -Uri 0.0.0.0:8000
    + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebExc
       eption
        + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand

[The configuration was introduced to facilitate testing from within Windows](https://github.com/w3c/web-platform-tests/pull/2673), but I haven't been able to reproduce the problem it was intended to solve. @domenic: I can change the desktop wallpaper on a Windows machine, but that's about it. Would you mind weighing in on this?

To be clear: the domain is referenced by just 7 tests, and by chance, none of them are invalidated by this behavior in GNU/Linux. I've included some information on each test below, but I'll need someone else to check things out on Windows. I'm also curious to know the situation on macOS; @gsnedders would you mind taking a look?

Two solutions come to mind:

- Remove that entry from the hosts file, assume the domain does not exist, and add tests in `infrastructure/assumptions/`
- Update the `create-hosts-file` command to emit the entry only when run from Windows

<details>
  <summary>Background on the existing references to the domain</summary>

Across all of web-platform-tests, only 11 files reference `nonexistent-origin`:

    $ git grep -l nonexistent-origin .
    acid/acid3/reference.sub.html
    content-security-policy/base-uri/report-uri-does-not-respect-base-uri.sub.html
    docs/_running-tests/chrome_android.md
    docs/_writing-tests/server-features.md
    tools/serve/serve.py
    webrtc/RTCPeerConnection-getIdentityAssertion.html
    webrtc/identity-helper.js
    xhr/event-error-order.sub.html
    xhr/event-error.sub.html
    xhr/send-network-error-async-events.sub.htm
    xhr/send-network-error-sync-events.sub.htm

Of these references, 3 are documentation and 1 is configuration. The remaining 7 are tests:

- `/xhr/event-error-order.sub.html`
  - Intention: produce a failed XHR request
  - Expectation: an HTTP 404 error for resource
  - Actual: an HTTP 404 error for CORS pre-flight request
- `/xhr/event-error.sub.html`
  - Intention: produce a failed XHR request
  - Expectation: an HTTP 404 error for resource
  - Actual: an HTTP 404 error for CORS pre-flight request
- `/xhr/send-network-error-async-events.sub.htm`
  - Intention: produce a failed XHR request
  - Expectation: an HTTP 404 error for resource
  - Actual: an HTTP 404 error for CORS pre-flight request
- `/xhr/send-network-error-sync-events.sub.htm`
  - Intention: produce a failed XHR request
  - Expectation: an HTTP 404 error for resource (because resource is not found)
  - Actual: an HTTP 404 error for CORS pre-flight (because method is not
    available\*)
- `/acid/acid3/reference.sub.html`
  - Intention: unclear
  - Expectation: unclear
  - Actual: a favicon is not assigned to the reference document
- `/webrtc/RTCPeerConnection-getIdentityAssertion.html`
  - Intention: Promise rejected with RTCError
  - Expectation: Promise rejected with RTCError
  - Actual: Test failure--RTCError is not defined in any implementation
- `/content-security-policy/base-uri/report-uri-does-not-respect-base-uri.sub.html`
  - Intention: ensure that the base URL is not referenced
  - Expectation: tests will fail if they make requests to a non-existent domain
  - Actual: tests may pass if they make requests to the domain

In other words: only the final test is impacted by this bug. For the other six tests, the bug violates the authors' expectations, but extenuating circumstances satisfy the underlying criteria.

In the case of the final test, the domain is being used to implicitly assert that some behavior (CSP report redirection) is *not* occurring. The bug subverts that check. This does not make the test invalid; it only makes it weaker than intended.

\* - the WPT server should probably respond with 405 in this case

</details>

**Comment by gsnedders** [comment]
If I'm not mistaken there's no comparable IP address on any other platform… (Which causes problems if there's some DNS server in use that rewrites everything to a search page…)

I guess we should only emit the entry on Windows, and add an assumptions test that it doesn't exist?

**Comment by jugglinmike** [comment]
Sounds good to me gh-10878
